### PR TITLE
Allow arbitrary parameter expressions

### PR DESCRIPTION
**Pull request by f0k**
First stab at #11, to continue the discussion based on actual code.

This gets the basic case working:
```
In [3]: l1 = lasagne.layers.DenseLayer((100, 10), 20)

In [4]: l2 = lasagne.layers.DenseLayer(l1, 10, W=l1.W.T)

In [5]: lasagne.layers.get_all_params(l2)
Out[5]: [b, W, b]

In [6]: l1.params

Out[6]: OrderedDict([(W, set(['trainable', 'regularizable'])), (b, set(['trainable']))])

In [7]: l2.params
Out[7]: OrderedDict([(W.T, set(['trainable', 'regularizable'])), (b, set(['trainable']))])
```
It will work just as fine with `T.exp(l1.W)`, something adding up two shared variables, or something involving no shared variables at all (e.g., a constant weight matrix).

If there are no objections to the way it's implemented, I can go ahead and update the documentation and tests, as well as the recurrent layers that now have a specific switch for supporting arbitrary expressions for `hid_init`.

Todo:
* [x] update documentation and tests
* [x] add helpful error message to `lasagne.updates.*` in case you're not passing shared variables only
* ~~update the recurrent layer to not distinguish between parameter expressions and parameter variables in the constructor, because the base class supports both now~~
* [x] update all the docstrings of all the layers because the parameters can be expressions now

### CODE REVIEW

**Comment by f0k** [comment]
Bump... anybody willing to have a look?

> If there are no objections to the way it's implemented, I can go ahead and update the documentation and tests, as well as the recurrent layers that now have a specific switch for supporting arbitrary expressions for `hid_init`.

**Comment by benanne** [comment]
Yes, sorry! It's on my to do list, but I'm moving this weekend. Will try to get around to it ASAP.

**Comment by skaae** [comment]
I can take a look in the evening.

**Comment by f0k** [comment]
Ah, all right! Don't worry, just wanted to make sure it wasn't forgotten. Good luck with moving!

**Comment by skaae** [comment]
I think this looks good.

For RNNs we could consider allowing `hid_init` be an input layer like mask? Then you could easily construct encoder/decoders without calling `get_output` multiple times.

Its not a huge issue though.

File: lasagne/utils.py [hunk 1]
```diff
@@ -76,6 +76,29 @@ def as_theano_expression(input):
...
+    Parameters
+    ----------
+    expressions : Theano expression or iterable of Theano expressions
+        The expressions to collect shared variables from.
+
+    Returns
+    -------
+    list of Theano shared variables
+        All shared variables the given expression(s) depend on, in the order
+        found by a left-recursive depth-first search. If some expressions are
```
Comment (benanne) [inline lasagne/utils.py]: Is there any particular reason you went with this search strategy? Just asking because it's mentioned explicitly. I guess it's just the easiest to implement?

File: lasagne/utils.py [hunk 2]
```diff
@@ -236,11 +259,11 @@ def create_param(spec, shape, name=None):
 
     Parameters
     ----------
-    spec : numpy array, Theano shared variable, or callable
+    spec : numpy array, Theano expression, or callable
         Either of the following:
 
         * a numpy array with the initial parameter values
-        * a Theano shared variable representing the parameters
+        * a Theano expression representing the parameters
```
Comment (benanne) [inline lasagne/utils.py]: I think that's a good idea, since that is the most common use case. Maybe something like "usually this is a Theano shared variable."

File: lasagne/layers/base.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ def get_params(self, **tags):
             result = [param for param in result
                       if not (self.params[param] & exclude)]
 
-        return result
+        return utils.collect_shared_vars(result)
```
Comment (benanne) [inline lasagne/layers/base.py]: I like how this is a one-line change now :)

**Comment by benanne** [comment]
LGTM! Just some comments, but nothing that needs to change as far as I can see. This is probably something we need to test thoroughly though, to ensure that we don't break any existing use cases.

**Comment by benanne** [comment]
Oh, three more things I thought of:

- the doc string for `add_param()` needs updating to indicate that expressions are now also accepted (but I guess you were planning on getting to that anyway).

- regarding parameter 'tagging', will everything still behave correctly when the keys for the tag dictionary can be arbitrary expressions? Are those hashable? I suppose so?

- in the issue I suggested that we should take precautions to ensure that people get meaningful error messages when they accidentally pass expressions where shared variables are expected (e.g. in `lasagne.updates`). Any concrete ideas on how to implement that without littering multiline checks all over the code?

File: lasagne/utils.py (hunk 1 above)
Comment (f0k) [inline lasagne/utils.py]: > I guess it's just the strategy that theano.gof.graph.inputs() implements

Yes, exactly, I copied it from there. I thought we should mention that there's a particular order. I can add that it's in topological order, like: "All shared variables the given expression(s) depend on, in topological order (as found by a left-recursive depth-first search)."

File: lasagne/layers/base.py (hunk 3 above)
Comment (f0k) [inline lasagne/layers/base.py]: (Although we could also change line 81 above to not listify `result`, for unmeasurably improved performance. Maybe I'll still do that.)

Comment (benanne) [inline lasagne/layers/base.py]: It's probably more readable as it is :)

File: lasagne/utils.py (hunk 1 above)
Comment (benanne) [inline lasagne/utils.py]: sure!

**Comment by f0k** [comment]
> the doc string for `add_param()` needs updating

Yes, and the `get_params()` docstrings should probably mention what happens for parameter expressions -- for sure that needs to be explained *somewhere*.

> will everything still behave correctly when the keys for the tag dictionary can be arbitrary expressions? Are those hashable? I suppose so?

Yes, they're hashable, theano needs that internally.

> in the issue I suggested that we should take precautions to ensure that people get meaningful error messages when they accidentally pass expressions where shared variables are expected (e.g. in `lasagne.updates`). Any concrete ideas on how to implement that without littering multiline checks all over the code?

Not really. There could be a utility function like `ensure_all_shared()` for the check that also raises the exception, but we'd still need to call it in every `lasagne.updates.*` function. It seems that's the only place we need it, though.

**Comment by benanne** [comment]
>  It seems that's the only place we need it, though, so we could also include it in get_or_compute_grads() that's called by every lasagne.updates.* function already.

Sounds like a plan :)

File: lasagne/utils.py (hunk 1 above)
Comment (f0k) [inline lasagne/utils.py]: I take that back, every order is a topological order since shared variables obviously don't have any parents :) I'll just leave it as it is.

**Comment by f0k** [comment]
Updated documentation and tests. Had to add a `reversed()` in `collect_shared_vars()` to make it respect the order of input expressions (this is due to how the underlying Theano graph traversal routine is implemented, and the `reverse` doesn't hurt as much as implementing the traversal routine ourselves).

Todo:
* [ ] add helpful error message to `lasagne.updates.*` in case you're not passing shared variables only
* [ ] update the recurrent layer to not distinguish between parameter expressions and parameter variables in the constructor, because the base class supports both now
* [ ] update all the docstrings of all the layers because the parameters can be expressions now

**Comment by benanne** [comment]
Looking good! When this is merged I'm thinking about writing a brief blog post about some of the possibilities this will offer (low-rank parameters, non-negativity constraints on parameters, easy parameter sharing etc.). Perhaps a bunch of examples like this would be useful in the docs as well.

**Comment by f0k** [comment]
Sorry for the lack of progress, I'm experimenting and didn't need it so far :)

> I'm thinking about writing a brief blog post

Sure, count me in if you'd like a co-author.

By the way, low-rank dense layers work really well with just popping a linear, biasless dense layer in front of the normal dense layer. [MEAN-NORMALIZED STOCHASTIC GRADIENT FOR LARGE-SCALE DEEP LEARNING](http://www.mirlab.org/conference_papers/International_Conference/ICASSP%202014/papers/p180-wiesler.pdf) made me a bit scared that you'd need fancy learning methods or batch normalization for that, but I found it to work just fine with sgd+nesterov.

> Perhaps a bunch of examples like this would be useful in the docs as well.

Somewhere I suggested that we'd move the existing `Layers` and `Creating custom layers` into a `Library walk-through` chapter, and add more pages to that covering the updates, regularization, and whatnot. We could also add parameter sharing and other tricks there! It's a lot of prose to write, though.

**Comment by benanne** [comment]
Yeah, I don't think I'll have time for that anytime soon myself. There's already a brief section on parameter sharing ( http://lasagne.readthedocs.org/en/latest/user/layers.html#parameter-sharing ), I guess we could rename and extend that.

**Comment by f0k** [comment]
> I guess we could rename and extend that.

Ah, yes, that would be a good spot for now!

**Comment by f0k** [comment]
Update:

1. `lasagne.updates.get_or_compute_grads()` now checks whether all params are shared variables. I'm not completely happy with this, but adding another function called in every `lasagne.updates.*` function seemed stupid as well. Note that with this change you can't use `get_or_compute_grads()` to compute gradients for things that are not shared variables, and that you can't use `sgd()` or anything else to compute updates for things that are not shared variables (but except for `sgd()`, all of them expect shared variables anyway, as they use `get_value()` to check the shape for creating accumulators).
2. The recurrent layers bypassed `create_param()` if `hid_init` or `cell_init` was a `TensorVariable`, and expected it to include the batch dimension if and only if it was a `TensorVariable`. Now that `create_param()` accepts any Theano expressions, I figured a better way to distinguish the two cases was via the dimensionality -- if it has enough dimensions to include the batch dimension, it probably includes the batch dimension, and if it has one dimension less, it will be replicated via a dot product against a vector of ones (as before). This required some more changes:
   (a) `create_param()` currently cannot be told to accept either a parameter of `(num_units)` or of `(batchsize, num_units)`. I've solved this by calling it a second time with the second shape if the first call failes (for which I changed `create_param()` to raise a `ValueError`, not a `RuntimeError`, before these were mixed).
   (b) `batchsize` is not usually known in a recurrent layer, only `num_units` is. I've solved this by changing `create_param()` to also accept `None` in its shape specification.
   Note that all this makes the recurrent layers incompatible with previous versions, because in previous versions `hid_init` was initialized to have a batch size of 1 for some reason, while now I initialize it to not include the batch size at all (otherwise we can't use `ndim` to know whether it needs replication over the batch or not).
  And again, I'm not completely happy with this. Maybe it would be nicer to change `create_param` to accept a set of alternative shapes that are also acceptable? The original, primary shape would be passed to a callable if a callable was given, and it would not be allowed to contain `None`s, but all other shapes may be underspecified and are only used to check whether a given shared variable, Theano expression or numpy array is accepted.

Any input or questions are welcome!

**Comment by f0k** [comment]
> By the way, low-rank dense layers work really well with just popping a linear, biasless dense layer in front of the normal dense layer.

A bit off-topic here, but I just saw that the authors of [DeepFont](http://arxiv.org/abs/1507.03196) train a low-rank dense layer by replacing it with a low-rank approximation after every update (Equation 3 on page 6). Fun.

**Comment by benanne** [comment]
It's interesting that training low-rank layers directly seems to be difficult with the current methods. It bothers me that the parameterization affects the efficiency of learning so much, but I guess that's the way it is. In speech, training a big fully connected layer and then replacing it with a low-rank version is quite popular as well (they call this a 'bottleneck', which gives you 'bottleneck features').

**Comment by benanne** [comment]
1. I think that's a good trade-off. It's the best place to put it, and it'll probably prevent some long-winded bug hunts in the near future :)

2. Since I haven't used the recurrent layers I find it hard to reason about the implications of this, so I'd like it if @craffel or @skaae (or someone else who knows the codebase) comments on this.

**Comment by f0k** [comment]
> I think that's a good trade-off. It's the best place to put it, and it'll probably prevent some long-winded bug hunts in the near future :)

Maybe we should solve #310 in Theano's `theano.updates.OrderedUpdates` and then use that one in `lasagne.updates`. It doesn't accept anything else but shared variables as keys, so that would remove the need of artificially constraining `get_or_compute_grads()`.

> Since I haven't used the recurrent layers I find it hard to reason about the implications of this

There's a general design question involved as well: To allow a layer to support different shapes for a particular parameter, we either need to extend `create_params()` to support multiple shapes, or have it called multiple times. What's better?
And to have it support shapes that are not completely known at construction time, we also need to extend it to allow `None`s in the specification. That's mostly useful in combination with symbolic parameter expressions, though, not so much with shared variables and not at all with weight initializers from `lasagne.init`. Again, would it be better to support this for secondary shapes only?

> so I'd like it if @craffel or @skaae (or someone else who knows the codebase) comments on this

Me too! @craffel, @skaae, what do you think of my proposed change to `hid_init`, i.e., distinguishing the two cases (including the batch dimension or not) via the dimensionality?

**Comment by craffel** [comment]
Your changes look reasonable to me; using `hid_init` as a `TensorVariable` is kind of an edge-case as it is.  AFAIK, @skaae is the only one using this.

**Comment by f0k** [comment]
> Your changes look reasonable to me; using `hid_init` as a `TensorVariable` is kind of an edge-case as it is.

Sure, but the change now requires the standard case to have `hid_init` be one dimension smaller than before (instead of adding a `1` as the first dimension -- we should have thought about that earlier). This will break compatibility with saved parameters of existing models. The alternative would be adding another keyword argument that tells whether `hid_init` and/or `cell_init` shall include the batch size or not, but there are so many keyword arguments already...

**Comment by craffel** [comment]
Yes, like I said, I don't think breaking compatibility is bad because I don't know of anyone using this except @skaae. It's an uncommon use-case.  He should probably give the go-ahead though.

**Comment by f0k** [comment]
> Yes, like I said, I don't think breaking compatibility is bad because I don't know of anyone using this except @skaae.

Maybe you're still misreading me -- this will break compatibility for everyone *except* @skaae.

**Comment by craffel** [comment]
Oops, I see what you're saying now, sorry.  Is there any reason not to require the hidden state to always be 2D, where the first dimension is either 1 (meaning "use the same initialization for all samples in the batch", which is the use-case the majority of the time), or `n_batch` (meaning "use a different initialization for each sample in the batch", arguably an edge-case)?  In the case of `n_batch = 1`, if we dotted it against a matrix of shape `(n_batch, 1)`, it would be harmless.  Sorry if you discussed this above and I missed it.

**Comment by f0k** [comment]
> Is there any reason not to require the hidden state to always be 2D, where the first dimension is either 1 (meaning "use the same initialization for all samples in the batch", which is the use-case the majority of the time), or `n_batch` (meaning "use a different initialization for each sample in the batch", arguably an edge-case)?

Yes, we cannot distinguish these two cases, unless we add an extra keyword to the constructor. All we can reliably check for a given `hid_init` is `ndim`. Maybe an extra constructor keyword would be better to distinguish the cases. If we add this, we could use it to exactly specify which axes the initial values are to be shared over -- e.g., for convolutional recurrent layers, one may want to have a single value per channel, not per spatial position. (That won't play nicely with the dotting trick any more, though.) Hmm, this PR is getting out of hands... maybe I should leave the recurrent layer untouched for now?

**Comment by craffel** [comment]
> Yes, we cannot distinguish these two cases, unless we add an extra keyword to the constructor. All we can reliably check for a given hid_init is ndim. Maybe an extra constructor keyword would be better to distinguish the cases. If we add this, we could use it to exactly specify which axes the initial values are to be shared over -- e.g., for convolutional recurrent layers, one may want to have a single value per channel, not per spatial position. (That won't play nicely with the dotting trick any more, though.) Hmm, this PR is getting out of hands... maybe I should leave the recurrent layer untouched for now?

Oh right, because we won't know whether to dot against a ones matrix in `get_output` unless we know the shape from the get-go, which we won't in general if `hid_init` is a `TensorVariable`.  I am personally ok with an extra kwarg.  Having lots of kwargs is not a big deal to me, as long as their defaults are sensible and their name/docstring is understandable.

**Comment by f0k** [comment]
Took out the changes to recurrent layers, they were getting too much to tackle in one go. Changed all relevant layer docstrings to accept expressions as an alternative to shared variables, numpy arrays or callables. Fixed some documentation errors on the way.

File: lasagne/layers/dense.py [hunk 4]
```diff
@@ -115,19 +115,18 @@ class NINLayer(Layer):
...
         dimension beyond the 2nd.
 
-    W : Theano shared variable, numpy array or callable
-        An initializer for the weights of the layer. If a shared variable or a
-        numpy array is provided the shape should be (num_inputs, num_units),
-        where num_units is the size of the 2nd. dimension of the input.
+    W : Theano shared variable, expression, numpy array or callable
+        Initial value, expression or initializer for the weights.
+        These should be a matrix with shape ``(num_inputs, num_units)``,
+        where ``num_inputs`` is the size of the second dimension of the input.
```
Comment (f0k) [inline lasagne/layers/dense.py]: > num_units is the size of the 2nd. dimension of the input

This was wrong, wasn't it?

Comment (benanne) [inline lasagne/layers/dense.py]: in what sense? That period definitely shouldn't have been there.

Comment (f0k) [inline lasagne/layers/dense.py]: > in what sense?

`num_units` vs. `num_inputs`

Comment (benanne) [inline lasagne/layers/dense.py]: oh.... right :)

**Comment by f0k** [comment]
Any other comments? Is this good to merge?

**Comment by benanne** [comment]
I went through the diff just now and everything looks good to me. I didn't bother to check the entire code base for docstrings you may have forgotten to update, but that wouldn't really be a showstopper anyway. Goot to merge as far as I'm concerned!

**Comment by f0k** [comment]
Okay, thanks! Will merge in the afternoon if there are no more comments.
